### DISCUSSION TITLE
Perf regression on v0.38.0 for HMR on Storybook + Vite

### DISCUSSION THREAD

**Discussion by adbutterfield** [chakra-ui/panda#2608] [General]
### Description

I noticed that HMR gets slower and slower after each change on Storybook after upgrading from v0.37.2. Versions from v.0.38.0 on seem to all have the issue.

I didn't check if it affects vite only, or also webpack.

### Link to Reproduction

Sorry, no reproduction 

### Steps to reproduce

1. You need a component library set up with Storybook + Vite
2. Install panda v0.38.0
3. Run `DEBUG="vite:plugin-transform" storybook dev`
4. Make a change that triggers HMR (component code or story code)
5. Check the time it takes to transform css in the console output
6. Repeat steps 4 & 5, the time it takes to transform css will continue to increase with each change (in my case starts 200-300ms, eventually going up to 4+ seconds)
7. Install panda v0.37.2
8. Repeat steps 3-6, time it takes to transform css stays at 200-300ms

### JS Framework

React

### Panda CSS Version

0.38.0

### Operating System

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by adbutterfield** [comment]
Should have wrote the description more like, "quickly goes up to 4+ seconds, and keeps going up".

**Comment by adbutterfield** [comment]
I think I found the root cause. I ran [clinic flame](https://clinicjs.org/flame/) on my storybook to see what code is taking so long to finish. Turns out it is `postcss-discard-duplicates` which was updated to v7 about a week ago.

The top 5 hottest blocks are from `postcss-discard-duplicates` running latest panda version. I don't get anything like that running v0.37.2.

Flame graph using v0.39.0
<img width="1440" alt="Screenshot 2024-05-20 at 10 27 13" src="https://github.com/chakra-ui/panda/assets/4752047/dbbb589c-67e0-4db4-8a50-a31f14956c3e">

Flame graph using where I removed `postcss-discard-duplicates` from `optimizePostCss` function
<img width="1440" alt="Screenshot 2024-05-20 at 10 58 51" src="https://github.com/chakra-ui/panda/assets/4752047/acc6d4a3-8d5e-4e2a-86d4-62a61a30cf32">

**Comment by adbutterfield** [comment]
It looks like `postcss-discard-duplicates` version didn't change. So maybe panda is creating more duplicates so `postcss-discard-duplicates` needs to do extra processing?

**Comment by adbutterfield** [comment]
I disabled `postcss-discard-duplicates`, and it looks like panda is generating more duplicates.

Initial load:
<img width="479" alt="Screenshot 2024-05-20 at 11 13 35" src="https://github.com/chakra-ui/panda/assets/4752047/74e84a75-2cd4-42f2-9b58-ffcc7b757079">

After one HMR:
<img width="480" alt="Screenshot 2024-05-20 at 11 13 53" src="https://github.com/chakra-ui/panda/assets/4752047/3b6c2443-3ffd-49af-8070-4400a14197f8">

After several more changes and HMR:
<img width="483" alt="Screenshot 2024-05-20 at 11 14 46" src="https://github.com/chakra-ui/panda/assets/4752047/2554fcb7-903e-45b5-9890-c051e0534001">

**Comment by adbutterfield** [comment]
It looks like a bug in the `CacheMap` introduced in `v0.38.0`, specifically where it's used for `atomic_cache`. If I change [this line](https://github.com/chakra-ui/panda/blob/2f63a4c093ed2e2f4405fb6bde6c069d93ce44ab/packages/core/src/style-decoder.ts#L32) to use a normal `Map`, the issue with duplicated styles goes away.

**Reply by adbutterfield** [reply]
Looking a bit deeper. I put a log on [this line](https://github.com/chakra-ui/panda/blob/2f63a4c093ed2e2f4405fb6bde6c069d93ce44ab/packages/core/src/layers.ts#L93). 
For each HMR, `utilities.nodes.length` gets (`encoder.atomic.length` - 1) * 3 bigger.

**Comment by frogdevvn** [comment]
Hmm. It seem i got the issue same with you with HMR and i using Vite... HMR slower after each changes and after restart Vite it normally but slower again

**Comment by astahmer** [comment]
hey, I wanted to look into this issue but couldnt reproduce it, can you send a minimal repro ?
I'm sure this happens on your side but it seems it might not happen for all setup using Panda (tried on the super simple [sandbox/vite](https://github.com/chakra-ui/panda/tree/main/sandbox/vite-ts), on the [sandbox/storybook](https://github.com/chakra-ui/panda/tree/main/sandbox/storybook) and on one private company repo with a component library)

btw thank you for the investigation, yes it's most likely the CacheMap, I made a [PR](https://github.com/chakra-ui/panda/pull/2612) that removes it in the Utility class so that Panda can properly de-duplicate styles

**Reply by adbutterfield** [reply]
So I figured it out. The bug only appears when you exceed `maxCacheSize`. Set it to something ridiculously small, like 5, and you can reproduce.

You can reproduce on this repo: https://github.com/adbutterfield/panda-turborepo

**Comment by segunadebayo** [comment]
This was fixed in [#2612](https://github.com/chakra-ui/panda/pull/2612) — the `CacheMap` (LRU cache) that was causing duplicate style accumulation in the decoder's `atomic_cache` was removed. The fix shipped in v0.39.0.

If anyone is still experiencing HMR slowdowns, please open a new issue with a reproduction.
